lc-0278-first-bad-version/src/main/java/solution/Solution.java

```diff
@@ -0,0 +1,37 @@
+package solution;
+
+public class Solution extends VersionControl {
+
+    Solution(int badVersion) {
+        super(badVersion);
+    }
+
+    // Perform binary search:
+    public int firstBadVersion(int n) {
+        int low = 1;
+        int high = n;
+
+        long mid = (low + high) / 2; // declare as long to handle math but 'mid' will always be within 'int'
+
+        while (low < high) {
+            mid = ((long) low + (long) high) / 2;
+            if (isBadVersion((int) mid)) {
```

Comment:
In my opinion, this:
```
if (isBadVersion((int) mid)) {
```
is less clear than:
```
if ((int)mid >= version >= mBadVersion) {
```

The method `isBadVersion` abstracts the comparison, which could be a good thing. However, you end up mixing two different levels of abstraction. The surrounding code is dealing with math, but this one line is hidden inside of an abstraction. It's actually very useful to see the raw comparsion here to undestand why you're updating the `high` or the `low`.

And, because `isBadVersion` is not used anywhere else, you never benefit from the abstraction. `isBadVersion` should be removed.

Reply:
I agree with this but the problem statement revolves around this abstraction [.](https://leetcode.com/problems/first-bad-version/). The problem simulates having API method `isBadVersion` which only returns true/false and I need to access this method the least amount of times to figure out the first bad version.

I have only implemented `VersionControl` into `Solution` to simulate the Leetcode structure and to have the solution method `firstBadVersion` match 1:1 to what I upload to Leetcode.